### PR TITLE
Backport lockstep changes to gazebo9

### PR DESCRIPTION
**Pull request by iche033**
~~This is WIP.~~  All changes have been backported from `gazebo11` branch and lockstep should be working. 

~~No ABI fixes have been made yet~~. Changes made. Testing ABI compatibility

~~Tests have not been backported yet too~~. Test backported. 

Creating a PR to trigger an ABI checker build so we can get a better idea of ABI issues that need to be addressed.

### CODE REVIEW

**Comment by iche033** [comment]
There are 3 ABI issues remaining:
https://build.osrfoundation.org/job/gazebo-abichecker-any_to_any-ubuntu_auto-amd64/405/API_5fABI_20report/

They are in `Noise.hh`, `Camera.hh`, `PhysicsTypes.hh`. I only added a new type in `PhysicsTypes.hh` and did not touch the remaining two headers. I'm wondering if they are false positives. Can I get a second opinion on this? cc @mabelzhang, @chapulina 

**Update:**

bd38258 fixed the 3 ABI iusses. Build is now blue.

**Comment by iche033** [comment]
This PR is ready for review.

File: test/integration/laser.cc [hunk 1]
```diff
@@ -608,6 +616,59 @@ TEST_P(LaserTest, LaserNoise)
   LaserUnitNoise(GetParam());
 }
 
+void LaserTest::LaserStrictUpdateRate(const std::string &_physicsEngine)
```
Comment (mabelzhang) [inline test/integration/laser.cc]: There is one warning on CI about unused parameter `_physicsEngine`. I must have missed it in gazebo11.

**Review by mabelzhang — APPROVED** [review]
I ran all the new tests and manually tested with camera_strict_rate.world to verify the frame rate as in gz11. Everything works.

CI tests are passing. There's just one warning in laser.cc.

File: test/integration/laser.cc (hunk 1 above)
Comment (iche033) [reply test/integration/laser.cc]: removed unused param in 95855f8

**Comment by mabelzhang** [comment]
Thank you for backporting everything! I already approved it. Feel free to merge!
I made the small fix for the warning for Gazebo 11 too in #2799. Could you please review that one? Thanks!

File: test/worlds/camera_strict_rate.world [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    <model name="box">
+      <pose>13 0 0.5 0 0 0</pose>
+      <link name="link">
+        <visual name="visual">
+          <geometry>
+            <box>
+              <size>1 1 1</size>
+            </box>
+          </geometry>
+        </visual>
```
Comment (mabelzhang) [inline test/worlds/camera_strict_rate.world]: I found in #2802 that these lines should be removed, to move the comparison camera to the pose where it should be.

```suggestion
```

Comment (iche033) [reply test/worlds/camera_strict_rate.world]: ported fix in a479501a5a2477d8154a3286e562dd5d2f431adc

**Pull request merged**